### PR TITLE
Add pass to optimize masked loads and stores.

### PR DESCRIPTION
**Pull request by ienkovich**
This patch adds a few patterns that can optimize out masks in a typical tiled loop of the following form:
```
  for i in range(0, tl.cdiv(size, TILE_SIZE)):
    offs = i * TILE_SIZE + tl.arange(0, TILE_SIZE)
    mask = offs < size
    val = tl.load(in_ptr + offs, mask=mask, other=0)
    res = ...
    tl.sotre(out_ptr + offs, res, mask=mask)
```
given a hint that `size` is divisible by `TILE_SIZE`. The first couple of patterns try to remove the cdiv-mul pair and transform the loop into something like:
```
  for i in range(0, size, TILE_SIZE):
    offs = i + tl.arange(0, TILE_SIZE)
    mask = offs < size
    val = tl.load(in_ptr + offs, mask=mask, other=0)
    res = ...
    tl.sotre(out_ptr + offs, res, mask=mask)
```
The last one tries to prove that the maximum possible value of `offs` is always less than `size`. If `size` is divisible by `TILE_SIZE` then `i <= size - TILE_SIZE` and `offs <= size - TILE_SIZE + TILE_SIZE - 1`, so we get `size - offs >= 1` which means the mask is always all-ones.

Example of TTIR before the optimization pass:
```
module {
  tt.func public @kernel(%arg0: !tt.ptr<f32> {tt.divisibility = 16 : i32} loc(unknown), %arg1: !tt.ptr<f32> {tt.divisibility = 16 : i32} loc(unknown), %arg2: i32 {tt.divisibility = 16 : i32} loc(unknown)) attrib\
utes {noinline = false} {
    %c0 = arith.constant 0 : index loc(#loc)
    %cst = arith.constant dense<[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]> : vector<16xi32> loc(#loc)
    %c15_i32 = arith.constant 15 : i32 loc(#loc)
    %c16_i32 = arith.constant 16 : i32 loc(#loc)
    %c0_i32 = arith.constant 0 : i32 loc(#loc)
    %c1_i32 = arith.constant 1 : i32 loc(#loc)
    %cst_0 = arith.constant dense<0.000000e+00> : vector<16xf32> loc(#loc)
    %0 = arith.addi %arg2, %c15_i32 : i32 loc(#loc)
    %1 = arith.divsi %0, %c16_i32 : i32 loc(#loc)
    %2 = vector.splat %arg2 : vector<16xi32> loc(#loc)
    scf.for %arg3 = %c0_i32 to %1 step %c1_i32  : i32 {
      %3 = arith.muli %arg3, %c16_i32 : i32 loc(#loc)
      %4 = vector.splat %3 : vector<16xi32> loc(#loc)
      %5 = arith.addi %4, %cst : vector<16xi32> loc(#loc)
      %6 = arith.cmpi slt, %5, %2 : vector<16xi32> loc(#loc)
      %7 = tt.addptr %arg0, %3 : !tt.ptr<f32>, i32 loc(#loc)
      %8 = triton_cpu.ptr_to_memref %7 : <f32> -> memref<16xf32> loc(#loc)
      %9 = vector.maskedload %8[%c0], %6, %cst_0 : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32> loc(#loc)
      %10 = tt.addptr %arg1, %3 : !tt.ptr<f32>, i32 loc(#loc)
      %11 = triton_cpu.ptr_to_memref %10 : <f32> -> memref<16xf32> loc(#loc)
      vector.maskedstore %11[%c0], %6, %9 : memref<16xf32>, vector<16xi1>, vector<16xf32> loc(#loc)
    } loc(#loc)
    tt.return loc(#loc)
  } loc(#loc)
} loc(#loc)
```
After the pass:
```
module {
  tt.func public @kernel(%arg0: !tt.ptr<f32> {tt.divisibility = 16 : i32} loc(unknown), %arg1: !tt.ptr<f32> {tt.divisibility = 16 : i32} loc(unknown), %arg2: i32 {tt.divisibility = 16 : i32} loc(unknown)) attrib\
utes {noinline = false} {
    %c16_i32 = arith.constant 16 : i32 loc(#loc)
    %c0 = arith.constant 0 : index loc(#loc)
    %c0_i32 = arith.constant 0 : i32 loc(#loc)
    scf.for %arg3 = %c0_i32 to %arg2 step %c16_i32  : i32 {
      %0 = tt.addptr %arg0, %arg3 : !tt.ptr<f32>, i32 loc(#loc)
      %1 = triton_cpu.ptr_to_memref %0 : <f32> -> memref<16xf32> loc(#loc)
      %2 = vector.load %1[%c0] : memref<16xf32>, vector<16xf32> loc(#loc)
      %3 = tt.addptr %arg1, %arg3 : !tt.ptr<f32>, i32 loc(#loc)
      %4 = triton_cpu.ptr_to_memref %3 : <f32> -> memref<16xf32> loc(#loc)
      vector.store %2, %4[%c0] : memref<16xf32>, vector<16xf32> loc(#loc)
    } loc(#loc)
    tt.return loc(#loc)
  } loc(#loc)
} loc(#loc)
```

### CODE REVIEW

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    MLIRContext *context = &getContext();
+    ModuleOp mod = getOperation();
+
+    // TODO: This pass optimizes out masks applying a set of very strict
+    // patterns. We should use more generic range and divisibility analysis
+    // to cover more cases and remove dependency on other transformations.
+    RewritePatternSet patterns1(context);
+    patterns1.add<CdivToDiv>(context);
+    if (failed(mlir::applyPatternsAndFoldGreedily(mod, std::move(patterns1))))
+      return signalPassFailure();
```
Comment (int3) [inline third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:367]: why do we want to return here instead of going on to try the other patterns below?

**Review by int3 — COMMENTED** [review]
Thanks for all the detailed comments!

File: test/TritonCPU/optimize-masks.mlir [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// RUN: triton-opt %s -split-input-file -triton-cpu-optimize-masks | FileCheck %s
+
+// Convert strided masked loads to scalar loads.
+
+// CHECK-LABEL: @remove_masks_in_for_loop
+// CHECK:       %[[VAL:.+]] = vector.load {{.+}} : memref<16xf32>, vector<16xf32>
+// CHECK:       vector.store %[[VAL]], {{.+}} : memref<16xf32>, vector<16xf32>
+
+#loc = loc("/localdisk/ilyaenko/triton-cpu/python/test/unit/cpu/test_opt.py":44:0)
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply test/TritonCPU/optimize-masks.mlir:9]: It was unintentional. Removed.

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+  }
+  return false;
+}
+
+// Optimize cdiv pattern using divisibility hints. If value is known to be
+// divisible by N then we can transform
+//   (val + K - 1) / K
+// to
+//   val / K
+// if N % K == 0 and val is not negitive. Usually, we cannot prove value to be
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:64]: Fixed

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+      return failure();
+
+    int64_t lowerVal = cast<IntegerAttr>(lowerDef.getValue()).getInt();
+    if (lowerVal < 0)
+      return failure();
+
+    // TODO: This is a strong requirement. With more generic value range
+    // analysis we should be able to not rely on this transformation.
+    if (!iv.hasOneUse())
+      return failure();
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:174]: If `iv` is used both scaled and unscaled, then we cannot scale the upper bound and step. Better range analysis should just be able to handle muli operations, but right now it's not in place.

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    int64_t scaleVal = cast<IntegerAttr>(scaleDef.getValue()).getInt();
+    int64_t divisorVal = cast<IntegerAttr>(divRhsDef.getValue()).getInt();
+    if (scaleVal != divisorVal || !isAlwaysDivisible(divLhs, scaleVal) ||
+        lowerVal % scaleVal != 0)
+      return failure();
+
+    // Build new lower bound.
+    Value newLower = lower;
+    if (lowerVal != 0) {
+      auto insPoint = rewriter.saveInsertionPoint();
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:196]: Thanks! I think I don't need to restore it at all.

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    maxOffs = buildMinOrMaxExpr(maskDef.getLhs(), isSigned, true, symbolTable);
+    minLen = buildMinOrMaxExpr(maskDef.getRhs(), isSigned, false, symbolTable);
+  } else {
+    maxOffs = buildMinOrMaxExpr(maskDef.getRhs(), isSigned, true, symbolTable);
+    minLen = buildMinOrMaxExpr(maskDef.getLhs(), isSigned, false, symbolTable);
+  }
+
+  // The mask is all-ones if max offset is always less than min length.
+  auto diff = maxOffs - minLen;
+  if (diff.getKind() == AffineExprKind::Constant) {
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:312]: Fixed

**Comment by ienkovich** [comment]
@int3 Thanks for the review! I've pushed my fixes.

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp (hunk 1 above)
Comment (int3) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:363]: ah I see. Still, why are we breaking it up into 3 runs instead of applying all patterns in one `RewritePatternSet`? Is the folding between each pattern application important?

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (int3) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:174]: would be nice to include that in the comment :)

**Review by int3 — APPROVED** [review]
just a question about the RewritePatternSet, but otherwise lgtm!

**Comment by int3** [comment]
As an aside, I'm surprised that LLVM doesn't already have a range analysis in place. Or maybe it exists but not at the MLIR level?

**Comment by ienkovich** [comment]
> As an aside, I'm surprised that LLVM doesn't already have a range analysis in place. Or maybe it exists but not at the MLIR level?

This optimization is built around the divisibility hint we have for input args. This optimization cannot be applied without a hint that the input vector size is divisible by the tile size (and that's why the optimization doesn't work for tile sizes > 16, we just don't have such a divisibility hint).

**Comment by int3** [comment]
Right but I was wondering if LLVM had some kind of range analysis framework we could build upon if we wanted to make this optimization more general, or if we'd have to write one from scratch

But I think I found the answer to my question: https://mlir.llvm.org/doxygen/classmlir_1_1dataflow_1_1IntegerRangeAnalysis.html

**Comment by ienkovich** [comment]
> But I think I found the answer to my question: https://mlir.llvm.org/doxygen/classmlir_1_1dataflow_1_1IntegerRangeAnalysis.html

I saw this one, but it's for statically known ranges. In our case, iv range is basically the whole int range. The same is true for offset and size. So I couldn't see how this type of analysis could help.

File: third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+    arith::ConstantOp addCstDef;
+    Value addOtherVal;
+    if (addCstDef = addOpDef.getLhs().getDefiningOp<arith::ConstantOp>())
+      addOtherVal = addOpDef.getRhs();
+    else if (addCstDef = addOpDef.getRhs().getDefiningOp<arith::ConstantOp>())
+      addOtherVal = addOpDef.getLhs();
+    else
+      return failure();
+
+    int64_t divisorCst = cast<IntegerAttr>(divisorDef.getValue()).getInt();
```
Comment (int3) [inline third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:90]: I was running the inductor tests and hit this assertion. I guess divSI can apply to floats too? Repro:

```
import triton
import triton.language as tl
import torch


@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr0):
    xnumel: tl.constexpr = 100
    XBLOCK: tl.constexpr = 128
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x1 = (xindex // 10)
    x2 = xindex
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp2 = tl.load(in_ptr1 + (x2), xmask)
    tmp1 = tmp0.to(tl.float64)
    tmp3 = tmp1 + tmp2
    tl.store(out_ptr0 + (x2), tmp3, xmask)


arg0 = torch.randn([1, 10, 1], dtype=torch.float32)
arg1 = torch.randn([10, 10], dtype=torch.float64)
out = torch.empty([1, 10, 10], dtype=torch.float64)
triton_[(1, )](arg0, arg1, out)
```

Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUTransforms/OptimizeMasks.cpp:90]: Thanks for the report! It fails for a vector form of DivSI. I'll post a fix.
